File: homog_map.m
%% Superpixels homogeneity classification
function map = homog_map(im, segments, labels_homog, labels_heter)

% Homogeneous = 1 (white)
im_homog = im;
for i=1:length(labels_homog)
    [rowi, coli] = find(segments==labels_homog(i));
    for j=1:length(rowi)
        im_homog(rowi(j),coli(j),:) = 1;
    end
end
% Heterogeneous = 0.5 (gray)
for i=1:length(labels_heter)
    [rowi1, coli1] = find(segments==labels_heter(i));
    for j=1:length(rowi1)
        im_homog(rowi1(j),coli1(j),:) = 0.5;
    end
end
% mark boundaries
%[sx,sy]=vl_grad(double(segments), 'type', 'forward');
%s = find(sx | sy);
map = im_homog(:,:,[1 2 3]);

% black edge
% map(1:end,1,:) = 0;
% map(1:end,end,:) = 0;
% map(1,1:end,:) = 0;
% map(end,1:end,:) = 0;

%map([s s+numel(im(:,:,1)) s+2*numel(im(:,:,1))]) = 0;
end

File: utils/homog_map.py
"""Build a homogeneity map image.

Translates MATLAB ``homog_map.m``.
"""

from __future__ import annotations

import numpy as np


def homog_map(
    data: np.ndarray,
    segments: np.ndarray,
    labels_homog: np.ndarray,
    labels_heter: np.ndarray,
) -> np.ndarray:
    """Create a 3-channel RGB image showing superpixel homogeneity.

    Homogeneous superpixels are painted white (1.0) and heterogeneous
    superpixels are painted gray (0.5).

    Parameters
    ----------
    data : np.ndarray
        Hyperspectral cube ``(rows, cols, bands)``.  Only its spatial
        dimensions are used; band values are overwritten.
    segments : np.ndarray
        2-D integer label map ``(rows, cols)``.
    labels_homog : np.ndarray
        Labels classified as homogeneous.
    labels_heter : np.ndarray
        Labels classified as heterogeneous.

    Returns
    -------
    np.ndarray
        RGB image ``(rows, cols, 3)`` with dtype ``float64``.
    """
    rows, cols = segments.shape
    rgb = np.zeros((rows, cols, 3), dtype=float)

    for label in labels_homog:
        mask = segments == label
        rgb[mask] = 1.0  # white

    for label in labels_heter:
        mask = segments == label
        rgb[mask] = 0.5  # gray

    return rgb
